### PR TITLE
drivers: display: Added support for 2.7inch epaper display

### PR DESCRIPTION
**Pull request by DineshDK03**
Added new display driver ek79652

Signed-off-by: Dinesh Kumar K <dinesh@linumiz.com>

### CODE REVIEW

File: boards/shields/waveshare_epaper/waveshare_epaper_gdew027w3.overlay [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+		busy-gpios = <&arduino_header 13 GPIO_ACTIVE_LOW>;	/* D7 */
+		pwr = [03 00 2b 2b 09];
+		softstart = [07 07 17];
+		cdi = < 87>;
+                pwropt = [60 A5 89 A5 90 00 93 2A A0 A5 A1 00];
+                lut-vcom-dc = [
+                         00 0E 00 00 00 01
+                         00 00 00 00 00 00
+                         00 00 00 00 00 00
+                         00 00 00 00 00 00
```
Comment (mbolivar-nordic) [inline boards/shields/waveshare_epaper/waveshare_epaper_gdew027w3.overlay:27]: This file is mixing tabs and spaces for indentation, and has leading spaces. Please fix.

File: dts/bindings/display/gooddisplay,ek79652.yaml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    width:
+      type: int
+      required: true
+      description: Width in pixel of the panel driven by the controller
+
+    reset-gpios:
+      type: phandle-array
+      required: true
+      description: RESET pin.
+
```
Comment (mbolivar-nordic) [inline dts/bindings/display/gooddisplay,ek79652.yaml:25]: Please fix the multiline string formatting. See https://github.com/zephyrproject-rtos/zephyr/pull/36189.

File: samples/subsys/display/lvgl/src/main.c [hunk 3]
```diff
@@ -15,6 +15,8 @@
 #include <logging/log.h>
 LOG_MODULE_REGISTER(app);
 
+#undef GDEW027W3
```
Comment (mbolivar-nordic) [inline samples/subsys/display/lvgl/src/main.c:18]: What's going on here? This is undefined, so the `#ifdef GDEW027W3` below is dead code.

If you need to match on a particular piece of hardware, then I think we need to get around to removing CONFIG_LVGL_DISPLAY_DEV_NAME and using a devicetree node instead.

Then you can either:

1. match on its compatible with DT_NODE_HAS_COMPAT: https://docs.zephyrproject.org/latest/reference/devicetree/api.html#c.DT_NODE_HAS_COMPAT
2. add some other devicetree property that the driver or application can use to decide to use `display_refresh` instead of `display_blanking_off`

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Hi and thanks for the patch. A few comments about the devicetree from my side.

File: dts/bindings/display/gooddisplay,ek79652.yaml (hunk 2 above)
Comment (DineshDK03) [reply dts/bindings/display/gooddisplay,ek79652.yaml]: @mbolivar-nordic  fixed!

File: boards/shields/waveshare_epaper/waveshare_epaper_gdew027w3.overlay (hunk 1 above)
Comment (DineshDK03) [reply boards/shields/waveshare_epaper/waveshare_epaper_gdew027w3.overlay]: @mbolivar-nordic  Fixed!

File: samples/subsys/display/lvgl/src/main.c (hunk 3 above)
Comment (DineshDK03) [reply samples/subsys/display/lvgl/src/main.c]: @mbolivar-nordic  As suggested , Inorder match for this particular hardware, used DT_NODE_HAS_COMPAT.

File: include/drivers/display.h [hunk 4]
```diff
@@ -81,6 +81,16 @@ enum display_orientation {
 	DISPLAY_ORIENTATION_ROTATED_270,
 };
 
+/**
```
Comment (parthitce) [inline include/drivers/display.h:84]: split the API changes to another commit.

File: boards/shields/waveshare_epaper/waveshare_epaper_gdew027w3.overlay [hunk 5]
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2021 Linumiz
+ *
```
Comment (parthitce) [inline boards/shields/waveshare_epaper/waveshare_epaper_gdew027w3.overlay:3]: Add SPDX license.

File: boards/shields/waveshare_epaper/waveshare_epaper_gdew027w3.overlay [hunk 6]
```diff
@@ -0,0 +1,70 @@
+/*
```
Comment (parthitce) [inline boards/shields/waveshare_epaper/waveshare_epaper_gdew027w3.overlay:1]: possibly split the shield in another commit

**Review by parthitce — APPROVED** [review]

File: include/drivers/display.h (hunk 4 above)
Comment (DineshDK03) [reply include/drivers/display.h:84]: @parthitce Modified!

File: boards/shields/waveshare_epaper/waveshare_epaper_gdew027w3.overlay (hunk 5 above)
Comment (DineshDK03) [reply boards/shields/waveshare_epaper/waveshare_epaper_gdew027w3.overlay:3]: @DineshDK03  As suggested added SPDX license

File: boards/shields/waveshare_epaper/waveshare_epaper_gdew027w3.overlay (hunk 6 above)
Comment (DineshDK03) [reply boards/shields/waveshare_epaper/waveshare_epaper_gdew027w3.overlay:1]: @parthitce  splitted shield related files in another commit

File: samples/subsys/display/lvgl/src/main.c (hunk 3 above)
Comment (mbolivar-nordic) [reply samples/subsys/display/lvgl/src/main.c]: > If you need to match on a particular piece of hardware, then I think we need to get around to removing CONFIG_LVGL_DISPLAY_DEV_NAME

This part was important, can you please do that too?

What I mean is that Kconfig options which just specify a device name have been deprecate for a long time now that the devicetree.h API has been stabilized and the device model has been updated to use devicetree nodes.

The thing to do now is to just get the `struct device*` from a devicetree pointer, using things like `DEVICE_DT_GET`, `DEVICE_DT_GET_ANY`, or `DEVICE_DT_GET_ONE`.

File: drivers/display/ek79652.c [hunk 7]
```diff
@@ -0,0 +1,556 @@
...
+			if (spi_write_dt(&driver->config->bus, &buf_set)) {
+				return -EIO;
+			}
+		}
+	}
+
+	return 0;
+}
+
+
```
Comment (parthitce) [inline drivers/display/ek79652.c:106]: nitpick, 1 new line is fine.

File: drivers/display/ek79652.c [hunk 8]
```diff
@@ -0,0 +1,556 @@
...
+	if (ek79652_write_cmd(driver, EK79652_CMD_DRF, NULL, 0)) {
+		return -EIO;
+	}
+
+	k_sleep(K_MSEC(EK79652_BUSY_DELAY));
+
+	return 0;
+}
+
+
```
Comment (parthitce) [inline drivers/display/ek79652.c:140]: nitpick, ditto

File: drivers/display/ek79652.c [hunk 9]
```diff
@@ -0,0 +1,556 @@
...
+		return -EIO;
+	}
+
+	if (ek79652_write_cmd(driver, EK79652_CMD_BTST, ek79652_softstart,
+			     sizeof(ek79652_softstart))) {
+		return -EIO;
+	}
+
+	/* Power optimization */
+	#if DT_INST_NODE_HAS_PROP(0, pwropt)
```
Comment (parthitce) [inline drivers/display/ek79652.c:403]: nitpick, no indentation

File: drivers/display/ek79652.c [hunk 10]
```diff
@@ -0,0 +1,556 @@
...
+
+	return 0;
+}
+
+static int ek79652_init(const struct device *dev)
+{
+	const struct ek79652_config *config = dev->config;
+	struct ek79652_data *driver = dev->data;
+
+	LOG_DBG("");
```
Comment (parthitce) [inline drivers/display/ek79652.c:493]: nitpick, can be removed

File: samples/subsys/display/lvgl/src/main.c [hunk 11]
```diff
@@ -48,7 +48,12 @@ void main(void)
 	lv_obj_align(count_label, NULL, LV_ALIGN_IN_BOTTOM_MID, 0, 0);
 
 	lv_task_handler();
+
+	#if DT_NODE_HAS_COMPAT(DT_INST(0, gooddisplay_ek79652), gooddisplay_ek79652)
```
Comment (parthitce) [inline samples/subsys/display/lvgl/src/main.c:52]: nitpick, no indentation

**Review by parthitce — DISMISSED** [review]

File: samples/subsys/display/lvgl/src/main.c (hunk 3 above)
Comment (DineshDK03) [reply samples/subsys/display/lvgl/src/main.c]: Apologies for the delayed reply,  if we remove `CONFIG_LVGL_DISPLAY_DEV_NAME` and get the device using any of the DT api , it will then go into hardware specific, also this [type](https://docs.zephyrproject.org/2.6.0/boards/shields/waveshare_epaper/doc/index.html#programming) of west build may create conflicts right?

File: drivers/display/ek79652.c (hunk 7 above)
Comment (DineshDK03) [reply drivers/display/ek79652.c:106]: fixed this and all the below chnages requested!

**Review by erwango — DISMISSED** [review]
+1 on shield part

File: samples/subsys/display/lvgl/src/main.c (hunk 3 above)
Comment (mbolivar-nordic) [reply samples/subsys/display/lvgl/src/main.c]: Thanks for the reply, but I do not understand. What conflicts are you referring to?

Comment (DineshDK03) [reply samples/subsys/display/lvgl/src/main.c]: I thought, if --DSHIELD overlay and device name used inside main.c will create conflicts but it wont. 
[here](https://github.com/zephyrproject-rtos/zephyr/pull/38306/commits/8208ed2d97b60ceb37474c9e2cf9db34e2c525f5) modified the requested changes

File: include/drivers/display.h [hunk 12]
```diff
@@ -81,6 +81,16 @@ enum display_orientation {
...
+ * @enum refresh_val
+ * @brief Enumeration with possible display refresh
+ *
+ */
+enum refresh_val {
+	PARTIAL_REFRESH = 0,
```
Comment (jfischer-no) [inline include/drivers/display.h:90]: PARTIAL_REFRESH what? panel? controller RAM? both? What do you mean by partial? Is this change really necessary? None of the subsystems can deal with it, special handling of this driver is not desired. The same functionality can be achieved by using blanking_off, https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/display/gd7965.c#L120.

File: drivers/display/ek79652.c [hunk 13]
```diff
@@ -0,0 +1,552 @@
...
+			buf.buf = &data[i];
+			buf.len = 1;
+			gpio_pin_set(driver->dc, EK79652_DC_PIN, 0);
+			if (spi_write_dt(&driver->config->bus, &buf_set)) {
+				return -EIO;
+			}
+		}
+	}
+
+	if (disp_data != NULL) {
```
Comment (jfischer-no) [inline drivers/display/ek79652.c:92]: Why so cumbersome? Why not just _ek79652_write_cmd()_, as in gd7965 driver from which you have taken over so many lines.

File: drivers/display/ek79652.c [hunk 14]
```diff
@@ -0,0 +1,552 @@
...
+	if (ek79652_write_cmd(driver, EK79652_CMD_DRF, NULL, 0)) {
+		return -EIO;
+	}
+
+	k_sleep(K_MSEC(EK79652_BUSY_DELAY));
+
+	return 0;
+}
+
+static int set_partial_lut(struct ek79652_data *driver)
```
Comment (jfischer-no) [inline drivers/display/ek79652.c:139]: These days, waveform settings should come from controller's OTP, best with consideration of temperature.

File: drivers/display/ek79652.c [hunk 15]
```diff
@@ -0,0 +1,552 @@
...
+
+	if (ek79652_write_cmd(driver, EK79652_CMD_LUTBB, lut_bb,
+			     sizeof(lut_bb))) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int ek79652_update_partial_display(struct ek79652_data *driver, uint8_t *data, size_t len)
```
Comment (jfischer-no) [inline drivers/display/ek79652.c:179]: Change it here and overall where it fits to
```
static int ek79652_update_partial_display(const struct device *dev,
                                           uint8_t *data, size_t len)
{
        struct ek79652_data *data = dev->data;
```

File: drivers/display/ek79652.c [hunk 16]
```diff
@@ -0,0 +1,552 @@
...
+#define EK79652_NUMOF_PAGES		(EPD_PANEL_WIDTH / \
+					 EK79652_PIXELS_PER_BYTE)
+#define EK79652_BUFFER_SIZE		(EK79652_NUMOF_PAGES * EPD_PANEL_HEIGHT)
+#define EK79652_PANEL_FIRST_GATE	0U
+#define EK79652_PANEL_LAST_GATE		(EPD_PANEL_HEIGHT - 1)
+#define EK79652_PANEL_FIRST_PAGE	0U
+#define EK79652_PANEL_LAST_PAGE		(EK79652_NUMOF_PAGES - 1)
+
+struct ek79652_data {
+	const struct ek79652_config *config;
```
Comment (jfischer-no) [inline drivers/display/ek79652.c:50]: That is wrong, just as it is wrong in gd7965 driver (I will fix gd7965 later), see https://github.com/zephyrproject-rtos/zephyr/pull/39708/commits/63f48e70bb5aea39533c4f7bd6fcbe6db1e0ff2a (part of https://github.com/zephyrproject-rtos/zephyr/pull/39708)

File: drivers/display/ek79652.c [hunk 17]
```diff
@@ -0,0 +1,552 @@
...
+	.read = ek79652_read,
+	.get_framebuffer = ek79652_get_framebuffer,
+	.set_brightness = ek79652_set_brightness,
+	.set_contrast = ek79652_set_contrast,
+	.get_capabilities = ek79652_get_capabilities,
+	.set_pixel_format = ek79652_set_pixel_format,
+	.set_orientation = ek79652_set_orientation,
+};
+
+DEVICE_DT_INST_DEFINE(0, ek79652_init, NULL,
```
Comment (jfischer-no) [inline drivers/display/ek79652.c:549]: New driver require support for multiple instances.

File: drivers/display/ek79652.c [hunk 18]
```diff
@@ -0,0 +1,552 @@
...
+	}
+#endif
+
+	/* Reset DFV_EN */
+	tmp[0] = EK79652_PDRF_VAL;
+	if (ek79652_write_cmd(driver, EK79652_CMD_PDRF, tmp, 1)) {
+		return -EIO;
+	}
+
+	/* Turn on: booster, controller, regulators, and sensor. */
```
Comment (jfischer-no) [inline drivers/display/ek79652.c:451]: Does it have a temperature sensor like gd7965 controller? Does this controller have valid waveforms in OTP ROM (If not, return it to gooddisplay)? Do you understand the relationship between temperature sensor, OTP ROM and waveforms? And that what you do in line 477 does not fit together?

File: samples/subsys/display/lvgl/src/main.c [hunk 19]
```diff
@@ -48,7 +48,12 @@ void main(void)
 	lv_obj_align(count_label, NULL, LV_ALIGN_IN_BOTTOM_MID, 0, 0);
 
 	lv_task_handler();
+
+#if DT_NODE_HAS_COMPAT(DT_INST(0, gooddisplay_ek79652), gooddisplay_ek79652)
+	display_refresh(display_dev, PARTIAL_REFRESH);
```
Comment (jfischer-no) [inline samples/subsys/display/lvgl/src/main.c:104]: NAK, no special handling.

File: boards/arm/mimxrt1050_evk/Kconfig.defconfig [hunk 20]
```diff
@@ -53,9 +53,6 @@ endif # NETWORKING
 
 if LVGL
 
-config LVGL_DISPLAY_DEV_NAME
```
Comment (jfischer-no) [inline boards/arm/mimxrt1050_evk/Kconfig.defconfig:56]: Who deprecated it and when?

File: samples/subsys/display/lvgl/src/main.c [hunk 21]
```diff
@@ -15,15 +15,66 @@
 #include <logging/log.h>
 LOG_MODULE_REGISTER(app);
 
+#if DT_NODE_HAS_STATUS(DT_INST(0, gooddisplay_ek79652), okay)
+#define LVGL_DISPLAY_DEV_NAME DT_LABEL(DT_INST(0, gooddisplay_ek79652))
```
Comment (jfischer-no) [inline samples/subsys/display/lvgl/src/main.c:19]: No, you can not do it that way. We probably need a chosen property like "zephyr,display-controller".

**Review by jfischer-no — CHANGES_REQUESTED** [review]
This cannot go in like this, API change needs to be discussed and is actually not necessary, driver has many issues.

File: include/drivers/display.h (hunk 12 above)
Comment (DineshDK03) [reply include/drivers/display.h:90]: I was planning to use gd7965.c driver for this[ 2.7inch display ](https://www.waveshare.com/w/upload/2/2d/2.7inch-e-paper-Specification.pdf) . unfortunately it doesn't worked, So I created new driver based on gd7965.c . PARTIAL_REFRESH means display update happens in that defined particular area based on SRAM data and LUT. this new API is introduced because of this particular display's driver IC has 2 commands for updating the display as full and partial with 12h and 16h commands.

File: drivers/display/ek79652.c (hunk 13 above)
Comment (DineshDK03) [reply drivers/display/ek79652.c:92]: Because according to [datasheet](https://ibb.co/FDBjLKQ)  inbetween sending bytes CS pin should be made HIGH.  Since CS pin need to be made LOW only when sending data, this display is working only when sending byte by byte data. In zephyr we can't toggle CS pin. so only _ek79652_write_data()_ function is introduced.

File: drivers/display/ek79652.c (hunk 14 above)
Comment (DineshDK03) [reply drivers/display/ek79652.c:139]: when I tried with OTP there is heavy ghosting effect happened on the display and also display refresh happened at very slowly. then only opted for this LUT set ,that too is only for partial_display resfresh.

File: drivers/display/ek79652.c (hunk 18 above)
Comment (DineshDK03) [reply drivers/display/ek79652.c:451]: yes it does have. _set_partial_lut_ is effective only for partial display refresh.

File: boards/arm/mimxrt1050_evk/Kconfig.defconfig (hunk 20 above)
Comment (DineshDK03) [reply boards/arm/mimxrt1050_evk/Kconfig.defconfig:56]: [Here](https://github.com/zephyrproject-rtos/zephyr/pull/38306#discussion_r748644311) @mbolivar-nordic  mentioned that it was deprecated.

File: samples/subsys/display/lvgl/src/main.c (hunk 21 above)
Comment (DineshDK03) [reply samples/subsys/display/lvgl/src/main.c:19]: Okay , will make changes based on chosen property in the latter commit

File: boards/arm/mimxrt1050_evk/Kconfig.defconfig (hunk 20 above)
Comment (jfischer-no) [reply boards/arm/mimxrt1050_evk/Kconfig.defconfig:56]: @DineshDK03 That was meant in general, look at the changes in #40765, now option LVGL_DISPLAY_DEV_NAME obtain value from devicetree only and hopefully will be removed after some LVGL adaption code rework. You probably want to rebase on latest main branch.

Comment (DineshDK03) [reply boards/arm/mimxrt1050_evk/Kconfig.defconfig]: @jfischer-no  Updated the LVGL_DISPLAY_DEV_NAME changes along with some minor changes mentioned in the driver ek79652.c except refresh APIs. expecting your comments on refresh API's and changes based on refresh API.

File: drivers/display/ek79652.c [hunk 22]
```diff
@@ -0,0 +1,706 @@
...
+	int err;
+
+	if (!spi_is_ready(&config->bus)) {
+		LOG_ERR("SPI bus %s not ready", config->bus.bus->name);
+		return -ENODEV;
+	}
+
+	if (!device_is_ready(config->reset.port)) {
+		LOG_ERR("GPIO port for EK79652 reset is not ready");
+		return -EIO;
```
Comment (mbolivar-nordic) [inline drivers/display/ek79652.c:534]: Can you please use -ENODEV consistently when a device is not ready?

File: samples/subsys/display/lvgl/src/main.c [hunk 23]
```diff
@@ -47,7 +47,12 @@ void main(void)
 	lv_obj_align(count_label, NULL, LV_ALIGN_IN_BOTTOM_MID, 0, 0);
 
 	lv_task_handler();
-	display_blanking_off(display_dev);
+
+	if (IS_ENABLED(CONFIG_EK79652)) {
+		display_refresh(display_dev, PARTIAL_REFRESH);
```
Comment (mbolivar-nordic) [inline samples/subsys/display/lvgl/src/main.c:52]: This is basically still a device-specific ifdef inside what is supposed to be a generic sample. Such things are generally bad practice in zephyr. Either the device API for displays should be able to tell the user if the display_refresh() or display_blanking_off() API should be used for a particular device, or there should be some common init function here that does the right thing. Doing this in a device-specific way won't scale.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Happy new year!

Looks a lot better from the DT / driver model side. I have a couple of requests left.

I'd like someone who knows displays (@jfischer-no ?) to ACK the device-specific parts before approving.

File: drivers/display/ek79652.c (hunk 22 above)
Comment (DineshDK03) [reply drivers/display/ek79652.c:534]: will rectify in the following commit

File: samples/subsys/display/lvgl/src/main.c (hunk 23 above)
Comment (DineshDK03) [reply samples/subsys/display/lvgl/src/main.c:52]: yes, I am completely agree that not to use device-specific ifdef in generic sample. this can be rectified only if the new API chnages are reviewed, since this particular display is having different registers for display refresh(partial or full), we need user input to do the display refresh accordingly. this API is mainly introduced to avoid complete display refresh when only part of the display needs refresh, say if we are using number counter with static text message , we don't need entire display to refresh when only the part where number counter is displayed. @jfischer-no need inputs from your side.

Comment (mbolivar-nordic) [reply samples/subsys/display/lvgl/src/main.c:52]: > @jfischer-no need inputs from your side.

ping

File: drivers/display/ek79652.c [hunk 24]
```diff
@@ -0,0 +1,725 @@
...
+		return err;
+	}
+
+	err = spi_write_dt(&config->bus, &buf_set);
+	if (err < 0) {
+		return err;
+	}
+
+	if (data != NULL) {
+		for (int i = 0 ; i < len ; i++) {
```
Comment (jfischer-no) [inline drivers/display/ek79652.c:79]: This loop does not make any sense for me, why cannot you reuse `struct spi_buf buf` like gd7965 driver or define a new one `struct spi_buf foo = {.buf = data, .len = len};` ?

>Because according to [datasheet](https://ibb.co/FDBjLKQ) inbetween sending bytes CS pin should be made HIGH. Since CS pin need to be made LOW only when sending data, this display is working only when sending byte by byte data. In zephyr we can't toggle CS pin. so only ek79652_write_data() function is introduced.

I doubt that.

File: drivers/display/ek79652.c (hunk 22 above)
Comment (jfischer-no) [reply drivers/display/ek79652.c:553]: Stop to resolve unresolved conversations.

File: samples/subsys/display/lvgl/src/main.c (hunk 23 above)
Comment (jfischer-no) [reply samples/subsys/display/lvgl/src/main.c:51]: Again, NACK to API changes, strong NACK to this change in the sample.

File: drivers/display/ek79652.c [hunk 25]
```diff
@@ -0,0 +1,725 @@
...
+#include <drivers/display.h>
+#include <drivers/gpio.h>
+#include <drivers/spi.h>
+#include <sys/byteorder.h>
+
+#include "ek79652_regs.h"
+
+#define LOG_LEVEL CONFIG_DISPLAY_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(ek79652);
```
Comment (jfischer-no) [inline drivers/display/ek79652.c:22]: just `LOG_MODULE_REGISTER(ek79652, CONFIG_DISPLAY_LOG_LEVEL);`

File: drivers/display/ek79652.c [hunk 26]
```diff
@@ -0,0 +1,725 @@
...
+	if (err < 0) {
+		return err;
+	}
+
+	err = ek79652_write_cmd(dev, EK79652_CMD_PDTM2, ptd, buf_ptd_len);
+	if (err < 0) {
+		return err;
+	}
+
+	if (driver->disp_refresh == true) {
```
Comment (jfischer-no) [inline drivers/display/ek79652.c:272]: What is the reason to call ek79652_write() with `driver->disp_refresh` set to false?

File: boards/shields/waveshare_epaper/waveshare_epaper_gdew027w3.overlay [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+		width = <176>;
+		height = <264>;
+		dc-gpios = <&arduino_header 15 GPIO_ACTIVE_LOW>;	/* D9 */
+		reset-gpios = <&arduino_header 14 GPIO_ACTIVE_LOW>;	/* D8 */
+		busy-gpios = <&arduino_header 13 GPIO_ACTIVE_LOW>;	/* D7 */
+		pwr = [03 00 2b 2b 09];
+		softstart = [07 07 17];
+		cdi = <0x87>;
+		pwropt = [60 A5 89 A5 90 00 93 2A A0 A5 A1 00];
+		lut-vcom-dc = [
```
Comment (jfischer-no) [inline boards/shields/waveshare_epaper/waveshare_epaper_gdew027w3.overlay:30]: As this display has valid OTP settings it should just use it, so remove any lut-* arrays.
https://github.com/zephyrproject-rtos/zephyr/pull/38306#discussion_r756862641
https://github.com/zephyrproject-rtos/zephyr/pull/38306#discussion_r756349108

**Review by jfischer-no — CHANGES_REQUESTED** [review]
I would like to see this PR for the ek79652 in a simple form (load wave setting from OTP, no "partial" updates) without additional commits with questionable changes in API and samples.

**Comment by DineshDK03** [comment]
@jfischer-no fixed all the comments you mentioned, except this https://github.com/zephyrproject-rtos/zephyr/pull/38306#discussion_r834533983 , the below image will explain that issue.


SPI_buffer_combined
SPI_buffer_split

**Comment by jfischer-no** [comment]
@DineshDK03 Please take a look at the latest changes @andysan has made in https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/display/uc81xx.c and https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/display/ssd16xx.c, and try to align your driver with these. In the commit history, I see shield support before the driver, this should be the other way around.

**Comment by andysan** [comment]
> @DineshDK03 Please take a look at the latest changes @andysan has made in https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/display/uc81xx.c and https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/display/ssd16xx.c, and try to align your driver with these. In the commit history, I see shield support before the driver, this should be the other way around.

I would suggest going even further. As far as I can tell, the register layout of this chip is very close (almost identical) to the UltraChip EPD drivers. A quick look at the data sheet suggests that you can most likely just add a quirk structure for this device in UC81xx and it should mostly just work out of the box. You might want to have a look at https://github.com/zephyrproject-rtos/zephyr/pull/48383 if you go down that route.
